### ISSUE TITLE
trait_duplication_in_bounds breaks code 

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#9961]
### Summary

.

### Lint Name

trait_duplication_in_bounds

### Reproducer

I tried this code:

```rust
trait Proj {
    type S;
}

impl Proj for () {
    type S = ();
}

impl Proj for i32 {
    type S = i32;
}

trait Base<T> {
    fn is_base(&self);
}

trait Derived<B: Proj>: Base<B::S> + Base<()> {
    fn is_derived(&self);
}

fn f<P: Proj>(obj: &dyn Derived<P>) {
    obj.is_derived();
    Base::<P::S>::is_base(obj);
    Base::<()>::is_base(obj);
}

fn main() {
    let _x: fn(_) = f::<()>;
    let _x: fn(_) = f::<i32>;
}

```

I saw this happen:
`cargo clippy --fix   -- -Aclippy::all -Wclippy::trait_duplication_in_bounds`
```
after fixes were automatically applied the compiler reported errors within these files:

  * src/main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0277]: the trait bound `dyn Derived<P>: Base<()>` is not satisfied
  --> src/main.rs:24:25
   |
24 |     Base::<()>::is_base(obj);
   |     ------------------- ^^^ the trait `Base<()>` is not implemented for `dyn Derived<P>`
   |     |
   |     required by a bound introduced by this call
   |
help: consider introducing a `where` clause, but there might be an alternative better way to express this requirement
   |
21 | fn f<P: Proj>(obj: &dyn Derived<P>) where dyn Derived<P>: Base<()> {
   |                                     ++++++++++++++++++++++++++++++

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
Original diagnostics will follow.

warning: these bounds contain repeated elements
  --> src/main.rs:17:25
   |
17 | trait Derived<B: Proj>: Base<B::S> + Base<()> {
   |                         ^^^^^^^^^^^^^^^^^^^^^ help: try: `Base<B::S>`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#trait_duplication_in_bounds
   = note: requested on the command line with `-W clippy::trait-duplication-in-bounds`
```

I expected to see this happen:

### Version

```text
rustc 1.67.0-nightly (b3bc6bf31 2022-11-24)
binary: rustc
commit-hash: b3bc6bf31265ac10946a0832092dbcedf9b26805
commit-date: 2022-11-24
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```
